### PR TITLE
Bump styletron v5

### PR DESCRIPTION
**Pull request by chasestarr**
#### Description

- edits baseui `styled` wrapper to perform fewer operations outside of styletron-react. now it only applies a simple wrapper injecting the `$theme`
- previously `styled()` calls were not type-checked, this is the cause of most code-churn as well as moving props type def to function generic rather than typing argument to style function.
- changes prettier parser to flow
- edits babel config to enable [callable generic syntax](https://babeljs.io/docs/en/babel-preset-flow#all)

#### Scope

- [x] Major: Breaking Change

### CODE REVIEW

**Comment by chasestarr** [comment]
todos:
[ ] write migration notes
[ ] remove usage of `$ref`

File: src/styles/styled.js [hunk 1]
```diff
@@ -20,28 +22,26 @@ const wrapper = StyledComponent =>
...
-  // use withStyleDeep, otherwise use baseStyled
-  let styleFn = baseStyled;
-  if (args[0] && args[0].__STYLETRON__) {
-    styleFn = withStyleDeep;
-  }
-  // Also allow passing deep style overrides via $style prop
-  // Ex: <StyledDiv $style={{color: 'red'}} />
-  // Issue for supporting this natively in styletron:
-  // https://github.com/rtsao/styletron/issues/221
+  <Props>(string, (WithTheme<Props>) => StyleObject): StyletronComponent<Props>,
```
Comment (tajo) [inline src/styles/styled.js]: Could we set here something as `<Props={}>` or maybe even `<Props=any>` (default `any` could turn this into non-breaking change?). So you don't ever have to do this:

```js
styled<{}>()
```

File: src/styles/styled.js [hunk 2]
```diff
@@ -20,28 +22,26 @@ const wrapper = StyledComponent =>
...
   };
 
-const baseStyled = createStyled({wrapper, getInitialStyle, driver});
+type StyletronComponent<Props> = React.StatelessFunctionalComponent<Props> & {
+  // eslint-disable-next-line flowtype/no-weak-types
+  __STYLETRON__: any,
+};
+type WithTheme<Props> = {$theme: ThemeT} & Props;
+type StyleFn = {
+  (string, StyleObject): StyletronComponent<{}>,
```
Comment (tajo) [inline src/styles/styled.js]: Could we also keep allowing

```js
type StyleFn = {
  (React.Component, StyleObject): StyletronComponent<{}>
```

so it matches the `styled` function from Styletron?

**Comment by tajo** [comment]
Good job!

File: src/styles/styled.js (hunk 1 above)
Comment (chasestarr) [reply src/styles/styled.js]: i tried this out, but led to flow errors when combined with this function signature overload syntax. the angle brackets are [required](https://flow.org/en/docs/types/generics/#toc-adding-defaults-to-parameterized-generics) anyway, so i chose to not dig too deep on it

File: src/styles/styled.js (hunk 2 above)
Comment (chasestarr) [reply src/styles/styled.js]: I wrote a [comment](https://github.com/uber-web/baseui/pull/1241/files#diff-b919b7f3bd332ab6f0e816211d3f6561R35) below this line with thoughts on why I did not add this option. If you're interested, let's pair up so I can show you the strange inconsistencies I experienced

File: documentation-site/static/examples/input/with-tags.js [hunk 3]
```diff
@@ -28,13 +28,13 @@ class TagSelect extends React.Component {
   state = {value: '', tags: ['hello']};
   ref = React.createRef();
 
-  componentDidMount() {
-    this.ref.current.addEventListener('keydown', this.handleKeyDown);
-  }
+  // componentDidMount() {
+  //   this.ref.current.addEventListener('keydown', this.handleKeyDown);
```
Comment (gergelyke) [inline documentation-site/static/examples/input/with-tags.js]: do we want to remove these?

File: documentation-site/static/examples/layer/basic-tether.js [hunk 4]
```diff
@@ -54,10 +55,11 @@ export default class BasicTether extends React.Component {
   };
 
   render() {
+    console.log(this.popperRef1.current);
```
Comment (gergelyke) [inline documentation-site/static/examples/layer/basic-tether.js]:
```suggestion
```

File: documentation-site/static/examples/textarea/basic.js [hunk 5]
```diff
@@ -13,6 +13,6 @@ export default () => (
     <br />
     <Textarea size={SIZE.compact} error={true} placeholder="Placeholder" />
     <br />
-    <Textarea size={SIZE.compact} disabled placeholder="Disabled textarea" />
+    <Textarea size={SIZE.compact} disabled placeholder="Disabled textarea" /> */}
```
Comment (gergelyke) [inline documentation-site/static/examples/textarea/basic.js]: should we remove this too?

File: src/block/block.js [hunk 6]
```diff
@@ -146,9 +149,6 @@ function Block({
...
 }
 
-Block.defaultProps = {
-  overrides: {},
-  as: 'div',
-};
-
-export default Block;
+const BlockComponent = React.forwardRef<BlockPropsT, HTMLElement>(Block);
+BlockComponent.displayName = 'Block';
```
Comment (gergelyke) [inline src/block/block.js]: I think the styletron babel plugin adds the displaynames - should we remove this ?

Comment (chasestarr) [reply src/block/block.js]: it's a `forwardRef` component not `styled` =)

Comment (gergelyke) [reply src/block/block.js]: ah, I've missed that, thanks!

File: src/datepicker/__tests__/datepicker.test.js [hunk 7]
```diff
@@ -66,41 +66,41 @@ describe('Datepicker', () => {
...
-
-    renderedInput.props().onFocus();
-    expect(component).toHaveState('isOpen', true);
-
-    // isOpen state value is passed to Popover
-    component.update();
-    renderedPopover = component.find(Popover).first();
-    expect(renderedPopover.props().isOpen).toEqual(true);
-  });
+  // test('isOpen state change on down arrow keydown event', () => {
```
Comment (tajo) [inline src/datepicker/__tests__/datepicker.test.js]: `test.skip` might be better to surface those?

File: src/breadcrumbs/breadcrumbs.js [hunk 8]
```diff
@@ -22,12 +23,19 @@ export function BreadcrumbsRoot(props: {|...BreadcrumbsPropsT, ...LocaleT|}) {
...
-  const [Icon, baseIconProps] = getOverrides(overrides.Icon, StyledIcon);
+  const [Icon, baseIconProps] = getOverrides(overrides.Icon, ChevronRight);
   const [Separator, baseSeparatorProps] = getOverrides(
     overrides.Separator,
     StyledSeparator,
   );
 
+  const iconOverrides = mergeOverrides(
+    {Svg: {style: {verticalAlign: 'text-bottom'}}},
+    baseIconProps,
```
Comment (nadiia) [inline src/breadcrumbs/breadcrumbs.js]: Should it be `baseIconProps .overrides`?

File: src/styles/styled.js (hunk 2 above)
Comment (chasestarr) [reply src/styles/styled.js]: based on in-person discussions it would be ideal to support this, but i'm not sure how to properly type it. does anyone have ideas?

File: src/button/__tests__/button.test.js [hunk 9]
```diff
@@ -65,19 +65,4 @@ describe('Button Component', () => {
       target: 'foo',
     });
   });
-
-  test('internalOnClick prevents external onClick while loading', () => {
```
Comment (nadiia) [inline src/button/__tests__/button.test.js]: Not sure what this test is about, but should we still keep it as it tests an edge case?

File: src/datepicker/__tests__/datepicker.test.js (hunk 7 above)
Comment (chasestarr) [reply src/datepicker/__tests__/datepicker.test.js]: thank you - uncommented

File: src/accordion/types.js [hunk 10]
```diff
@@ -116,16 +115,8 @@ export type StatefulPanelPropsT = $Diff<
...
-  $as?: string,
-  // styled function wrapper related
-  $style?: ?{},
-  /* eslint-disable flowtype/no-weak-types */
-  $ref?: React.Ref<any>,
-};
-
-export type SharedStylePropsT = SharedStylePropsArgT & {
-  $theme: ThemeT,
+  $size?: string | number,
```
[the remark this answers is not in the document]
Comment (chasestarr) [reply src/accordion/types.js]: I made them optional types for this reason

File: src/checkbox/styled-components.js [hunk 11]
```diff
@@ -189,20 +191,20 @@ export const Checkmark = styled('span', props => {
     marginBottom: $theme.sizing.scale0,
     marginLeft: $theme.sizing.scale0,
     marginRight: $theme.sizing.scale0,
-  };
+  }: {});
```
Comment (nadiia) [inline src/checkbox/styled-components.js]: Is there a reason we need to type it here but not other styled fns?

File: src/datepicker/__tests__/datepicker.test.js [hunk 12]
```diff
@@ -113,18 +113,18 @@ describe('Datepicker', () => {
...
-    const component = mount(
-      // $FlowFixMe
-      <Datepicker onChange={onChange} range value={[]} />,
-    );
-
-    component.instance().onChange(data);
-    expect(onChange).toBeCalledWith(data);
-    expect(component).toHaveState('isOpen', true);
-  });
+  // test('does not set isOpen state to false if a single date from a range selected', () => {
```
Comment (nadiia) [inline src/datepicker/__tests__/datepicker.test.js]: Does it mean this functionality if broken now? We should create tickets for it and the one above.

File: src/datepicker/datepicker.js [hunk 13]
```diff
@@ -183,24 +183,26 @@ export default class Datepicker extends React.Component<
...
-                disabled={this.props.disabled}
-                value={this.formatDisplayValue(this.props.value)}
-                onFocus={this.open}
-                onBlur={this.handleInputBlur}
-                onKeyDown={this.handleKeyDown}
-                placeholder={this.props.placeholder}
-                required={this.props.required}
-                {...inputProps}
-              />
+              <span>
```
Comment (nadiia) [inline src/datepicker/datepicker.js]: What's the span wrapper is for?

File: src/input/__tests__/base-input.test.js [hunk 14]
```diff
@@ -103,18 +103,14 @@ test('BaseInput - autoFocus sets the initial focus state', () => {
...
     onChange: jest.fn(),
-    inputRef: {current: {focus}},
   };
 
   // $FlowFixMe
   const wrapper = mount(<BaseInput {...props} />);
   // Is focused when mount
   expect(wrapper).toHaveState('isFocused', true);
-  // ref's focus method is called
-  expect(focus).toBeCalled();
```
Comment (nadiia) [inline src/input/__tests__/base-input.test.js]: Why not keep this test to make sure the `autoFocus` works as expected?

File: src/popover/__tests__/popover.test.js [hunk 15]
```diff
@@ -132,10 +134,6 @@ describe('Popover', () => {
       onPopperUpdate: wrapperInstance.onPopperUpdate,
       placement: wrapper.state().placement,
     });
-    expect(tetherProps.anchorRef).toBe(wrapperInstance.anchorRef.current);
-    expect(tetherProps.arrowRef).toBe(wrapperInstance.arrowRef.current);
-    expect(tetherProps.popperRef).toBe(wrapperInstance.popperRef.current);
```
Comment (nadiia) [inline src/popover/__tests__/popover.test.js]: We should keep these tests to make sure it functions as expected.

File: src/popover/__tests__/popover.test.js [hunk 16]
```diff
@@ -254,45 +252,22 @@ describe('Popover', () => {
...
-      {
-        $ref: wrapper.instance().anchorRef,
-        onClick: wrapper.instance().onAnchorClick,
-        'aria-controls': 'bui-mock-id',
-        'aria-haspopup': 'true',
-        'aria-expanded': 'true',
-      },
-      {},
-    );
+    expect(wrapper).toMatchSnapshot();
```
Comment (nadiia) [inline src/popover/__tests__/popover.test.js]: I think explicit test cases have more value here to make sure it functions as expected.

**Comment by nadiia** [comment]
There are also some visual changes introduces to datepicker (its dropdown icons have incorrect color), pagination, side-navigation, button's left/right padding.

File: docs/migrations/01241.md [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+- type PropsT = {$active: boolean, $theme: ThemeT};
+- const MyStyledComponent = styled('div', (props: PropsT) => {
+-  return {
+-    backgroundColor: props.$active ? props.$theme.colors.primary400 : props.$theme.colors.negative400,
+-  };
+- });
+
++ type PropsT = {$active: boolean};
++ const MyStyledComponent = styled<PropsT>('div', props => {
```
Comment (kahwee) [inline docs/migrations/01241.md]: Specify the type of the object: `(props: PropsT)`

File: docs/migrations/01241.md [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
++  return {
++    backgroundColor: props.$active ? props.$theme.colors.primary400 : props.$theme.colors.negative400,
```
Comment (kahwee) [inline docs/migrations/01241.md]: Incorrect example? Looks like you still have `props.$theme`

File: docs/migrations/01241.md (hunk 17 above)
Comment (chasestarr) [reply docs/migrations/01241.md]: it's not necessary since `PropsT` is provided as a generic

File: docs/migrations/01241.md (hunk 18 above)
Comment (chasestarr) [reply docs/migrations/01241.md]: `ThemeT` will be typed in props now, without the need to explicitly define it. But what i did overlook is if you have custom fields in your theme. we probably need a way to pass a custom theme type as a generic and fallback to the default theme type

File: src/checkbox/styled-components.js (hunk 11 above)
Comment (chasestarr) [reply src/checkbox/styled-components.js]: it's because `null` is not valid [here](https://github.com/uber-web/baseui/pull/1241/files/1ab536dfc0376d74483df83699da7111ef1abff7#diff-5d4eec77dcb85451082356a170ef0e24R182). this coerces the type

File: src/datepicker/datepicker.js (hunk 13 above)
Comment (chasestarr) [reply src/datepicker/datepicker.js]: The `Input` component does not implement ref forwarding

File: src/input/masked-input.js [hunk 19]
```diff
@@ -21,15 +21,16 @@ type PropsT = {
   maskChar?: string,
 };
 
-function MaskOverride(props: PropsT) {
+const MaskOverride = React.forwardRef((props, ref) => {
+  const {inputRef, ...restProps} = props;
```
[the remark this answers is not in the document]
Comment (chasestarr) [reply src/input/masked-input.js]: correct, I want to exclude `inputRef` and pass forwarded ref
